# Segmenting and Clustering Neighborhoods in Toronto 
> We will explore neighborhoods in Toronto and group them into clusters.

- toc: true
- branch: master
- badges: true
- comments: true
- categories: [segmentation, k-means clustering, web scraping]



In this blog post, we will explore neighborhoods in Toronto, Canada.

First, we will build the code to scrape an HTML table from this Wikipedia page (https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M) that lists postal codes in Canada.  

We will need `requests` for getting the HTML contents of the website and `lxml.html` for parsing the relevant fields. We will then store this data on a pandas dataframe.

In [1]:
# import libraries
import requests
import lxml.html as lh
import pandas as pd
import numpy as np

---
### Scrape the HTML table cells.

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

# create a handle page to handle the contents of the website
page = requests.get(url)

# store the contents of the website under doc
doc = lh.fromstring(page.content)

# parse data that are stored between <tr>...</tr> of HTML
tr_elements = doc.xpath('//tr')

Ensure that all the rows have the same width. If not, we probably got something more that just the table.

In [3]:
# check the length of the first 10 rows
[len(T) for T in tr_elements[:10]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

This means that there are 3 columns per row.  
Now parse the table header.

In [4]:
# parse the first row as the header
tr_elements = doc.xpath('//tr')

# create an empty list
col = []
i = 0

# for each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name = t.text_content()
    print('%d:"%s"'%(i,name))
    col.append((name,[]))

1:"Postcode"
2:"Borough"
3:"Neighbourhood
"


---
### Create pandas dataframe
Each header is appended to a tuple along with an empty list

In [5]:
# since the first row is the header, data is stored on the second row onwards
for j in range(1, len(tr_elements)):
    #T is our j'th row
    T = tr_elements[j]
    
    # if row is not of size 3, the //tr data is not from our table 
    if len(T)!= 3:
        break
    
    # i is the index of our column
    i = 0
    
    # iterate through each element of the row
    for t in T.iterchildren():
        data = t.text_content() 
        # check if row is empty
        if i > 0:
        # convert any numerical value to integers
            try:
                data = int(data)
            except:
                pass
        # append the data to the empty list of the i'th column
        col[i][1].append(data)
        # increment i for the next column
        i+=1

Check the length of each column. Ideally, there should all be the same.

In [6]:
[len(C) for (title, C) in col]

[287, 287, 287]

This shows that each of the columns has exactly 287 rows.

Create the dataframe.

In [7]:
Dict = {title:column for (title, column) in col}
df = pd.DataFrame(Dict)
df.head()

Postcode           Borough     Neighbourhood\n
0      M1A      Not assigned      Not assigned\n
1      M2A      Not assigned      Not assigned\n
2      M3A        North York         Parkwoods\n
3      M4A        North York  Victoria Village\n
4      M5A  Downtown Toronto      Harbourfront\n

In [8]:
df.columns = ['Postcode', 'Borough', 'Neighborhood']
cols = df.columns.tolist()
cols

['Postcode', 'Borough', 'Neighborhood']

Clean the messy string in the Neighborhood column.

In [9]:
df = df.replace('\n', ' ', regex=True)
df.head()

Postcode           Borough       Neighborhood
0      M1A      Not assigned      Not assigned 
1      M2A      Not assigned      Not assigned 
2      M3A        North York         Parkwoods 
3      M4A        North York  Victoria Village 
4      M5A  Downtown Toronto      Harbourfront

Drop all cells with a Borough that is "Not assigned".

In [10]:
df.drop(df.index[df['Borough'] == 'Not assigned'], inplace=True)
# reset the index and drop the previous index
df = df.reset_index(drop=True)
df.head(10)

Postcode           Borough       Neighborhood
0      M3A        North York         Parkwoods 
1      M4A        North York  Victoria Village 
2      M5A  Downtown Toronto      Harbourfront 
3      M6A        North York  Lawrence Heights 
4      M6A        North York    Lawrence Manor 
5      M7A  Downtown Toronto      Queen's Park 
6      M9A         Etobicoke  Islington Avenue 
7      M1B       Scarborough             Rouge 
8      M1B       Scarborough           Malvern 
9      M3B        North York   Don Mills North

Combine neighborhoods based on similar Postcode and Borough.

In [11]:
df = df.groupby(['Postcode', 'Borough'])['Neighborhood'].apply(','.join).reset_index()
df.columns = ['Postcode', 'Borough', 'Neighborhood']
df.head(10)

Postcode      Borough                                      Neighborhood
0      M1B  Scarborough                                   Rouge ,Malvern 
1      M1C  Scarborough           Highland Creek ,Rouge Hill ,Port Union 
2      M1E  Scarborough                Guildwood ,Morningside ,West Hill 
3      M1G  Scarborough                                           Woburn 
4      M1H  Scarborough                                        Cedarbrae 
5      M1J  Scarborough                              Scarborough Village 
6      M1K  Scarborough      East Birchmount Park ,Ionview ,Kennedy Park 
7      M1L  Scarborough                  Clairlea ,Golden Mile ,Oakridge 
8      M1M  Scarborough  Cliffcrest ,Cliffside ,Scarborough Village West 
9      M1N  Scarborough                      Birch Cliff ,Cliffside West

Remove any spacing at the start of the string.

In [12]:
df['Neighborhood'] = df['Neighborhood'].str.strip()

Assign Borough values to the Neighborhood where value is "Not assigned".

In [13]:
df.loc[df['Neighborhood'] == 'Not assigned', 'Neighborhood'] = df['Borough']

In [14]:
# shape of the dataframe
df.shape

(103, 3)

In [15]:
# save this file as a csv
df.to_csv(r'df_can.csv')

---
### Get the coordinates of each neighborhood
Read the geospatial data into a pandas dataframe.

In [16]:
link = 'http://cocl.us/Geospatial_data'
df1 = pd.read_csv(link)
df1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [17]:
# shape of the data
df1.shape

(103, 3)

Change the column name 'Postal code' to 'Postcode' to merge the two data frames together.

In [18]:
df1.columns = ['Postcode', 'Latitude', 'Longitude']

cols = df1.columns.tolist()
cols

['Postcode', 'Latitude', 'Longitude']

Read the csv file saved in the previous assignment.

In [19]:
df = pd.read_csv('df_can.csv', index_col=0)
df.head()

Postcode      Borough                            Neighborhood
0      M1B  Scarborough                          Rouge ,Malvern
1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union
2      M1E  Scarborough       Guildwood ,Morningside ,West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

Merge the two dataframes togehter based on their Postcode.

In [20]:
df_new = pd.merge(df, df1, on='Postcode')
df_new.head()

Postcode      Borough                            Neighborhood   Latitude  \
0      M1B  Scarborough                          Rouge ,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union  43.784535   
2      M1E  Scarborough       Guildwood ,Morningside ,West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [21]:
# save the file as a csv
df_new.to_csv(r'df_final.csv')

---
### Segmentation and Clustering

In [22]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [23]:
neighborhoods = pd.read_csv('df_final.csv', index_col=0)
neighborhoods.head()

Postcode      Borough                            Neighborhood   Latitude  \
0      M1B  Scarborough                          Rouge ,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union  43.784535   
2      M1E  Scarborough       Guildwood ,Morningside ,West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

Find out how many boroughs and neighborhoods are in the dataset.

In [24]:
print('The dataframe has {} boroughs and {} neighborhoods'.format(
    len(neighborhoods['Borough'].unique()), neighborhoods.shape[0]
))

The dataframe has 10 boroughs and 103 neighborhoods


Use the geopy library to get the latitude and longitude values of Toronto, Canada.

In [25]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent='can_explorer')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinates of Toronto are {}, {}'.format(latitude, longitude))

The geographical coordinates of Toronto are 43.653963, -79.387207


---
### Visualize the neighborhoods.

In [26]:
# create a map of Toronto 
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Utilize the Foursquare API to explore the neighborhoods and segment them.

In [27]:
# limit the number of venues returned by the foursquare API
LIMIT = 100

# define Foursquare Credentials
CLIENT_ID = 'XXXX' 
CLIENT_SECRET = 'XXXX'
VERSION = '20191219'

---
### Explore neighborhoods in Toronto

Let's create a function to repeat the same process to all the neighborhoods in Toronto.

In [28]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Run the above function on each neighborhood and create a new dataframe called toronto_venues.

In [29]:
#collapse-output
toronto_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                 latitudes=neighborhoods['Latitude'],
                                 longitudes=neighborhoods['Longitude'])

Rouge ,Malvern
Highland Creek ,Rouge Hill ,Port Union
Guildwood ,Morningside ,West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park ,Ionview ,Kennedy Park
Clairlea ,Golden Mile ,Oakridge
Cliffcrest ,Cliffside ,Scarborough Village West
Birch Cliff ,Cliffside West
Dorset Park ,Scarborough Town Centre ,Wexford Heights
Maryvale ,Wexford
Agincourt
Clarks Corners ,Sullivan ,Tam O'Shanter
Agincourt North ,L'Amoreaux East ,Milliken ,Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview ,Henry Farm ,Oriole
Bayview Village
Silver Hills ,York Mills
Newtonbrook ,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park ,Don Mills South
Bathurst Manor ,Downsview North ,Wilson Heights
Northwood Park ,York University
CFB Toronto ,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens ,Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West ,

In [30]:
# glimpse of the new dataset
toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                          Rouge ,Malvern              43.806686   
1                          Rouge ,Malvern              43.806686   
2  Highland Creek ,Rouge Hill ,Port Union              43.784535   
3       Guildwood ,Morningside ,West Hill              43.763573   
4       Guildwood ,Morningside ,West Hill              43.763573   

   Neighborhood Longitude                  Venue  Venue Latitude  \
0              -79.194353                Wendy's       43.807448   
1              -79.194353  Interprovincial Group       43.805630   
2              -79.160497  Royal Canadian Legion       43.782533   
3              -79.188711      G & G Electronics       43.765309   
4              -79.188711             Marina Spa       43.766000   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.200378            Print Shop  
2       -79.163085                   Bar  
3       -79.191537     Electronics Store  
4       -79.191000                   Spa

In [31]:
# dimensions of the new datset
toronto_venues.shape

(2214, 7)

Let's check how many venues were returned for each neighborhood.

In [32]:
#collapse-output
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide ,King ,Richmond                                              100   
Agincourt                                                               4   
Agincourt North ,L'Amoreaux East ,Milliken ,Ste...                      3   
Albion Gardens ,Beaumond Heights ,Humbergate ,J...                     12   
Alderwood ,Long Branch                                                 10   
Bathurst Manor ,Downsview North ,Wilson Heights                        19   
Bayview Village                                                         4   
Bedford Park ,Lawrence Manor East                                      25   
Berczy Park                                                            55   
Birch Cliff ,Cliffside West                                             4   
Bloordale Gardens ,Eringate ,Markland Wood ,Old...                      8   
Brockton ,Exhibition Place ,Parkdale Village                           22   
Business Reply Mail Processing Centre 969 Eastern                      15   
CFB Toronto ,Downsview East                                             2   
CN Tower ,Bathurst Quay ,Island airport ,Harbou...                     17   
Cabbagetown ,St. James Town                                            42   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  12   
Cedarbrae                                                               8   
Central Bay Street                                                     79   
Chinatown ,Grange Park ,Kensington Market                              88   
Christie                                                               18   
Church and Wellesley                                                   86   
Clairlea ,Golden Mile ,Oakridge                                        10   
Clarks Corners ,Sullivan ,Tam O'Shanter                                12   
Cliffcrest ,Cliffside ,Scarborough Village West                         4   
Cloverdale ,Islington ,Martin Grove ,Princess G...                      1   
Commerce Court ,Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        9   
Deer Park ,Forest Hill SE ,Rathnelly ,South Hil...                     14   
Del Ray ,Keelesdale ,Mount Dennis ,Silverthorn                          4   
Design Exchange ,Toronto Dominion Centre                              100   
Don Mills North                                                         4   
Dorset Park ,Scarborough Town Centre ,Wexford H...                      7   
Dovercourt Village ,Dufferin                                           15   
Downsview ,North Park ,Upwood Park                                      4   
Downsview Central                                                       4   
Downsview Northwest                                                     4   
Downsview West                                                          6   
East Birchmount Park ,Ionview ,Kennedy Park                             4   
East Toronto                                                            4   
Emery ,Humberlea                                                        3   
Fairview ,Henry Farm ,Oriole                                           63   
First Canadian Place ,Underground city                                100   
Flemingdon Park ,Don Mills South                                       22   
Forest Hill North ,Forest Hill West                                     4   
Glencairn                                                               6   
Guildwood ,Morningside ,West Hill                                       7   
Harbord ,University of Toronto                                         35   
Harbourfront                                    

Let's find out how many unique categories can be curated from all the returned venues.

In [33]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix='', prefix_sep='')

# add neighborhood column back to the dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood']

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    1               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0        0            0                0         0   
2               0        0            0                0         0   
3               0        0            0                0         0   
4               0        0            0                0         0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Bus

In [34]:
# dimensions of the new dataframe
toronto_onehot.shape

(2214, 267)

Let's group rows by neighborhood by taking the mean of the frequency of occurence of each category.

In [35]:
#collapse-output
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                             Adelaide ,King ,Richmond     0.000000   
1                                            Agincourt     0.000000   
2    Agincourt North ,L'Amoreaux East ,Milliken ,St...     0.000000   
3    Albion Gardens ,Beaumond Heights ,Humbergate ,...     0.000000   
4                               Alderwood ,Long Branch     0.000000   
5      Bathurst Manor ,Downsview North ,Wilson Heights     0.000000   
6                                      Bayview Village     0.000000   
7                    Bedford Park ,Lawrence Manor East     0.000000   
8                                          Berczy Park     0.000000   
9                          Birch Cliff ,Cliffside West     0.000000   
10   Bloordale Gardens ,Eringate ,Markland Wood ,Ol...     0.000000   
11        Brockton ,Exhibition Place ,Parkdale Village     0.000000   
12   Business Reply Mail Processing Centre 969 Eastern     0.000000   
13                         CFB Toronto ,Downsview East     0.000000   
14   CN Tower ,Bathurst Quay ,Island airport ,Harbo...     0.000000   
15                         Cabbagetown ,St. James Town     0.000000   
16                                 Caledonia-Fairbanks     0.000000   
17               Canada Post Gateway Processing Centre     0.000000   
18                                           Cedarbrae     0.000000   
19                                  Central Bay Street     0.012658   
20           Chinatown ,Grange Park ,Kensington Market     0.000000   
21                                            Christie     0.000000   
22                                Church and Wellesley     0.011628   
23                     Clairlea ,Golden Mile ,Oakridge     0.000000   
24             Clarks Corners ,Sullivan ,Tam O'Shanter     0.000000   
25     Cliffcrest ,Cliffside ,Scarborough Village West     0.000000   
26   Cloverdale ,Islington ,Martin Grove ,Princess ...     0.000000   
27                      Commerce Court ,Victoria Hotel     0.000000   
28                                          Davisville     0.000000   
29                                    Davisville North     0.000000   
30   Deer Park ,Forest Hill SE ,Rathnelly ,South Hi...     0.000000   
31      Del Ray ,Keelesdale ,Mount Dennis ,Silverthorn     0.000000   
32            Design Exchange ,Toronto Dominion Centre     0.000000   
33                                     Don Mills North     0.000000   
34   Dorset Park ,Scarborough Town Centre ,Wexford ...     0.000000   
35                        Dovercourt Village ,Dufferin     0.000000   
36                  Downsview ,North Park ,Upwood Park     0.000000   
37                                   Downsview Central     0.000000   
38                                 Downsview Northwest     0.000000   
39                                      Downsview West     0.000000   
40         East Birchmount Park ,Ionview ,Kennedy Park     0.000000   
41                                        East Toronto     0.000000   
42                                    Emery ,Humberlea     0.000000   
43                        Fairview ,Henry Farm ,Oriole     0.000000   
44              First Canadian Place ,Underground city     0.000000   
45                    Flemingdon Park ,Don Mills South     0.000000   
46                 Forest Hill North ,Forest Hill West     0.000000   
47                                           Glencairn     0.000000   
48                   Guildwood ,Morningside ,West Hill     0.000000   
49                      Harbord ,University of Toronto     0.028571   
50                                        Harbourfront     0.000000   
51   Harbourfront East ,Toronto Islands ,Union Station     0.000000   
52                       High Park ,The Junction South     0.000000   
53              Highland Creek ,Rouge Hill ,Port Union     0.000000   
54                                   Hillcrest Village     0.000000   
55   Humber Bay ,King's Mill Park ,Kingsway Park So...     0.00000

In [36]:
# confirm the new size
toronto_grouped.shape

(101, 267)

Let's print the top 5 most common venues for each neighborhood.

In [37]:
#collapse-output
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide ,King ,Richmond----
              venue  freq
0       Coffee Shop  0.07
1        Restaurant  0.05
2              Café  0.04
3   Thai Restaurant  0.04
4  Sushi Restaurant  0.03


----Agincourt----
                       venue  freq
0                     Lounge  0.25
1  Latin American Restaurant  0.25
2             Breakfast Spot  0.25
3             Clothing Store  0.25
4              Metro Station  0.00


----Agincourt North ,L'Amoreaux East ,Milliken ,Steeles East----
         venue  freq
0         Park  0.33
1   Playground  0.33
2  Coffee Shop  0.33
3  Yoga Studio  0.00
4  Men's Store  0.00


----Albion Gardens ,Beaumond Heights ,Humbergate ,Jamestown ,Mount Olive ,Silverstone ,South Steeles ,Thistletown----
                 venue  freq
0        Grocery Store  0.17
1          Pizza Place  0.17
2           Beer Store  0.08
3  Fried Chicken Joint  0.08
4  Japanese Restaurant  0.08


----Alderwood ,Long Branch----
            venue  freq
0     Pizza Place   0.2
1        Phar

Let's put this data into a pandas dataframe.  
First, write a function to sort the venues in descending order.

In [38]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)

    return row_categories_sorted.index.values[0:num_top_venues]

Let's create the new dataframe and display the top 10 venues for each neighborhood.

In [39]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to the number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0                           Adelaide ,King ,Richmond   
1                                          Agincourt   
2  Agincourt North ,L'Amoreaux East ,Milliken ,St...   
3  Albion Gardens ,Beaumond Heights ,Humbergate ,...   
4                             Alderwood ,Long Branch   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                Coffee Shop            Restaurant                  Café   
1  Latin American Restaurant                Lounge        Breakfast Spot   
2                       Park           Coffee Shop            Playground   
3              Grocery Store           Pizza Place   Japanese Restaurant   
4                Pizza Place           Coffee Shop              Pharmacy   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0       Thai Restaurant            Steakhouse      Sushi Restaurant   
1        Clothing Store            Donut Shop                 Diner   
2               Dog Run      Department Store          Dessert Shop   
3          Liquor Store              Pharmacy        Discount Store   
4          Skating Rink                   Pub                  Pool   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                   Bar    Seafood Restaurant                Lounge   
1        Discount Store   Distribution Center               Dog Run   
2    Dim Sum Restaurant                 Diner        Discount Store   
3            Beer Store   Fried Chicken Joint        Sandwich Place   
4        Sandwich Place          Dance Studio                   Gym   

  10th Most Common Venue  
0         Cosmetics Shop  
1       Doner Restaurant  
2    Distribution Center  
3   Fast Food Restaurant  
4    Distribution Center

---
### Cluster the neighborhoods

Run the k-means algorithm to cluster the neighborhood into 5 clusters.

In [40]:
# set the number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# rin k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 1, 2, 0, 0, 0, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [41]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

Postcode      Borough                            Neighborhood   Latitude  \
0      M1B  Scarborough                          Rouge ,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union  43.784535   
2      M1E  Scarborough       Guildwood ,Morningside ,West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             0.0  Fast Food Restaurant            Print Shop   
1 -79.160497             2.0                   Bar         Women's Store   
2 -79.188711             0.0          Intersection        Breakfast Spot   
3 -79.216917             0.0           Coffee Shop     Korean Restaurant   
4 -79.239476             2.0   Fried Chicken Joint                Bakery   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Women's Store               Dog Run          Dessert Shop   
1      Doner Restaurant    Dim Sum Restaurant                 Diner   
2    Mexican Restaurant        Medical Center   Rental Car Location   
3     Indian Restaurant         Women's Store      Doner Restaurant   
4    Athletics & Sports       Thai Restaurant  Caribbean Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Dim Sum Restaurant                 Diner        Discount Store   
1        Discount Store   Distribution Center               Dog Run   
2     Electronics Store                   Spa            Donut Shop   
3                 Diner        Discount Store   Distribution Center   
4           Gas Station      Hakka Restaurant                  Bank   

  9th Most Common Venue 10th Most Common Venue  
0   Distribution Center       Doner Restaurant  
1            Donut Shop            Fabric Shop  
2      Doner Restaurant                Dog Run  
3               Dog Run              Drugstore  
4   Dumpling Restaurant              Drugstore

Let's visualize the clusters.

In [42]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        fill=True,
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

---
### Examine the clusters

You can examine and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can assign a name to each cluster.

**Cluster 1**

In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
0         Scarborough             0.0       Fast Food Restaurant   
2         Scarborough             0.0               Intersection   
3         Scarborough             0.0                Coffee Shop   
5         Scarborough             0.0                 Playground   
6         Scarborough             0.0           Department Store   
8         Scarborough             0.0                      Motel   
9         Scarborough             0.0            College Stadium   
10        Scarborough             0.0          Indian Restaurant   
12        Scarborough             0.0  Latin American Restaurant   
13        Scarborough             0.0                Pizza Place   
15        Scarborough             0.0       Fast Food Restaurant   
18         North York             0.0             Clothing Store   
19         North York             0.0         Chinese Restaurant   
21         North York             0.0                  Piano Bar   
22         North York             0.0           Ramen Restaurant   
24         North York             0.0              Grocery Store   
27         North York             0.0                Coffee Shop   
28         North York             0.0                Coffee Shop   
29         North York             0.0         Miscellaneous Shop   
32         North York             0.0          Korean Restaurant   
34         North York             0.0      Portuguese Restaurant   
36          East York             0.0                       Park   
37       East Toronto             0.0          Health Food Store   
38          East York             0.0                Coffee Shop   
39          East York             0.0          Indian Restaurant   
41       East Toronto             0.0           Greek Restaurant   
42       East Toronto             0.0             Sandwich Place   
43       East Toronto             0.0                       Café   
45    Central Toronto             0.0                       Park   
46    Central Toronto             0.0             Clothing Store   
47    Central Toronto             0.0               Dessert Shop   
49    Central Toronto             0.0                Coffee Shop   
51   Downtown Toronto             0.0                Coffee Shop   
52   Downtown Toronto             0.0                Coffee Shop   
53   Downtown Toronto             0.0                Coffee Shop   
54   Downtown Toronto             0.0                Coffee Shop   
55   Downtown Toronto             0.0                Coffee Shop   
56   Downtown Toronto             0.0                Coffee Shop   
57   Downtown Toronto             0.0                Coffee Shop   
58   Downtown Toronto             0.0                Coffee Shop   
59   Downtown Toronto             0.0                Coffee Shop   
60   Downtown Toronto             0.0                Coffee Shop   
61   Downtown Toronto             0.0                Coffee Shop   
62         North York             0.0                Coffee Shop   
63    Central Toronto             0.0             Ice Cream Shop   
65    Central Toronto             0.0                       Café   
66   Downtown Toronto             0.0                       Café   
67   Downtown Toronto             0.0                        Bar   
68   Downtown Toronto             0.0            Airport Service   
69   Downtown Toronto             0.0                Coffee Shop   
70   Downtown Toronto             0.0                Coffee Shop   
71         North York             0.0              Women's Store   
72         North York             0.0                Pizza Place   
73               York             0.0                      Trail   
75   Downtown Toronto             0.0              Grocery Store   
77       West Toronto             0.0                        Bar   
78       West Toronto             0.0                       Café   
80               York             0.0          Convenience Store   
82

**Cluster 2**

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough             1.0                  Park   
20        North York             1.0                  Park   
23        North York             1.0                  Park   
25        North York             1.0                  Park   
30        North York             1.0                  Park   
40         East York             1.0                  Park   
44   Central Toronto             1.0                  Park   
48   Central Toronto             1.0                  Park   
50  Downtown Toronto             1.0                  Park   
64   Central Toronto             1.0                  Park   
74              York             1.0                  Park   
90         Etobicoke             1.0                  Park   
98              York             1.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14           Coffee Shop            Playground               Dog Run   
20             Cafeteria    College Rec Center         Deli / Bodega   
23     Convenience Store                  Bank    Dim Sum Restaurant   
25              Bus Stop     Food & Drink Shop   Distribution Center   
30               Airport      Doner Restaurant          Dessert Shop   
40           Coffee Shop     Convenience Store               Dog Run   
44           Swim School              Bus Line               Dog Run   
48          Tennis Court            Playground   Distribution Center   
50                 Trail            Playground      Department Store   
64                 Trail         Jewelry Store      Sushi Restaurant   
74                Market         Women's Store             Gift Shop   
90                 River   Empanada Restaurant     Electronics Store   
98         Deli / Bodega   Empanada Restaurant     Electronics Store   

          5th Most Common Venue        6th Most Common Venue  \
14             Department Store                 Dessert Shop   
20            Electronics Store  Eastern European Restaurant   
23                        Diner               Discount Store   
25                 Dessert Shop           Dim Sum Restaurant   
30           Dim Sum Restaurant                        Diner   
40                 Dessert Shop           Dim Sum Restaurant   
44           Dim Sum Restaurant                        Diner   
48             Department Store                 Dessert Shop   
50                 Dessert Shop           Dim Sum Restaurant   
64                 Dessert Shop           Dim Sum Restaurant   
74            German Restaurant  Eastern European Restaurant   
90  Eastern European Restaurant          Dumpling Restaurant   
98  Eastern European Restaurant          Dumpling Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14    Dim Sum Restaurant                 Diner        Discount Store   
20   Dumpling Restaurant             Drugstore            Donut Shop   
23   Distribution Center               Dog Run      Doner Restaurant   
25                 Diner        Discount Store               Dog Run   
30        Discount Store   Distribution Center               Dog Run   
40                 Diner        Discount Store   Distribution Center   
44        Discount Store   Distribution Center      Doner Restaurant   
48    Dim Sum Restaurant                 Diner        Discount Store   
50                 Diner        Discount Store   Distribution Center   
64                 Diner        Discount Store   Distribution Center   
74   Dumpling Restaurant             Drugstore            Donut Shop   
90             Drugstore            Donut Shop  Ethiopian Restaurant   
98             Drugstore            Donut Shop      Doner Restaurant   

    10th Most Common Venue  
14     Distribution Center  
20        Doner Restaurant  
23        Department Store  
25    Ethiopian Restaurant  
30              Donut Shop  
40        Doner Restaurant  
44        Department Store

**Cluster 3**

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
1     Scarborough             2.0                        Bar   
4     Scarborough             2.0        Fried Chicken Joint   
7     Scarborough             2.0                     Bakery   
11    Scarborough             2.0  Middle Eastern Restaurant   
17     North York             2.0                Golf Course   
26     North York             2.0                       Café   
31     North York             2.0              Grocery Store   
33     North York             2.0              Grocery Store   
35      East York             2.0                Pizza Place   
76   West Toronto             2.0                   Pharmacy   
79     North York             2.0           Basketball Court   
81           York             2.0              Grocery Store   
101     Etobicoke             2.0              Grocery Store   
102     Etobicoke             2.0        Rental Car Location   

    2nd Most Common Venue     3rd Most Common Venue  \
1           Women's Store          Doner Restaurant   
4                  Bakery        Athletics & Sports   
7                Bus Line              Intersection   
11                 Bakery            Breakfast Spot   
17                Dog Run  Mediterranean Restaurant   
26   Gym / Fitness Center      Caribbean Restaurant   
31                   Park                      Bank   
33   Gym / Fitness Center              Liquor Store   
35           Intersection                 Pet Store   
76                 Bakery                      Park   
79                   Park                    Bakery   
81            Pizza Place                   Brewery   
101           Pizza Place       Japanese Restaurant   
102             Drugstore                       Bar   

          4th Most Common Venue 5th Most Common Venue  \
1            Dim Sum Restaurant                 Diner   
4               Thai Restaurant  Caribbean Restaurant   
7                          Park          Soccer Field   
11                   Smoke Shop               Dog Run   
17                         Pool   Dumpling Restaurant   
26          Japanese Restaurant      Doner Restaurant   
31                Shopping Mall                 Hotel   
33           Athletics & Sports         Women's Store   
35         Gym / Fitness Center              Pharmacy   
76                         Bank                  Café   
79   Construction & Landscaping         Women's Store   
81                     Bus Line         Women's Store   
101                Liquor Store              Pharmacy   
102                     Dog Run          Dessert Shop   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1               Discount Store   Distribution Center               Dog Run   
4                  Gas Station      Hakka Restaurant                  Bank   
7               Ice Cream Shop           Bus Station         Metro Station   
11          Dim Sum Restaurant                 Diner        Discount Store   
17                   Drugstore            Donut Shop      Doner Restaurant   
26          Dim Sum Restaurant                 Diner        Discount Store   
31        Colombian Restaurant          Dessert Shop     Electronics Store   
33            Doner Restaurant    Dim Sum Restaurant                 Diner   
35                   Gastropub                  Bank  Fast Food Restaurant   
76   Middle Eastern Restaurant                  Pool           Supermarket   
79                       Diner        Discount Store   Distribution Center   
81                     Dog Run    Dim Sum Restaurant                 Diner   
101             Discount Store            Beer Store   Fried Chicken Joint   
102         Dim Sum Restaurant                 Diner        Discount Store   

    9th Most Common Venue       10th Most Common Venue  
1              Donut Shop                  Fabric Shop  
4     Dumpling Restaurant                    Drugstore  
7           Women's Store

**Cluster 4**

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
94  Etobicoke             3.0          Home Service         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
94               Dog Run          Dessert Shop    Dim Sum Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
94                 Diner        Discount Store   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
94      Doner Restaurant          Deli / Bodega

**Cluster 5**

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
96  North York             4.0           Pizza Place         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
96               Dog Run      Department Store          Dessert Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
96    Dim Sum Restaurant                 Diner        Discount Store   

   9th Most Common Venue 10th Most Common Venue  
96   Distribution Center       Doner Restaurant

Cluster 1 is the biggest. Clusters 4 and 5 have just one neighborhood each.